### DISCUSSION TITLE
Fail2ban 1.0.2 with Exim-4.95_5 filter not worked

### DISCUSSION THREAD

**Discussion by ysha0** [fail2ban/fail2ban#3497] [General]
Hi

I'm using Fail2ban 1.0.2 (py39-fail2ban FreeBSD package) with Exim-4.95_5, but filter does not catch such attempts by spammers:

```
2023-04-01 07:34:55 SMTP call from [164.52.0.102] dropped: too many syntax or protocol errors (last command was "?\035?\027?\036?\031?\030?#???\026???\027???\r?0?.\004\003\005\003\006\003\b\007\b\b\b \b", NULL)
2023-04-01 07:34:55 SMTP call from [164.52.0.102] dropped: too many syntax or protocol errors (last command was "?\035?\027?\036?\031?\030?#???\026???\027???\r?0?.\004\003\005\003\006\003\b\007\b\b\b \b", NULL)
2023-04-01 07:35:18 SMTP call from [164.52.0.102] dropped: too many unrecognized commands (last was "\033%-12345X")
2023-04-01 07:35:18 SMTP call from [164.52.0.102] dropped: too many unrecognized commands (last was "Accept-Language: zh-CN,zh;q=0.8,en-US;q=0.5,en;q=0.3")
2023-04-01 07:35:32 SMTP call from [164.52.0.102] dropped: too many unrecognized commands (last was "(last was "{"id": 1, "method": "mining.hello", "params": {"agent":"ethminer-0.17","host":"213.183.101.138", "port":"19","proto":"EthereumStratum/2.0.0"}}")
2023-04-01 07:35:49 Connection from [164.52.0.102] refused: too many connections from that IP address

```
filter.d/exim.conf
```
# Fail2Ban filter for exim
#
# This includes the rejection messages of exim. For spam and filter
# related bans use the exim-spam.conf
#


[INCLUDES]

# Read common prefixes. If any customizations available -- read them from
# exim-common.local
before = exim-common.conf

[Definition]

# Fre-filter via "prefregex" is currently inactive because of too different failure syntax in exim-log (testing needed):
#prefregex = ^%(pid)s <F-CONTENT>\b(?:\w+ authenticator failed|([\w\-]+ )?SMTP (?:(?:call|connection) from|protocol(?: synchronization)? error)|no MAIL in|(?:%(host_info_pre)s\[[^\]]+\]%(host_info_suf)s(?:sender verify fail|rejected RCPT|dropped|AUTH command))).+</F-CONTENT>$

failregex = ^%(pid)s %(host_info)ssender verify fail for <\S+>: (?:Unknown user|Unrouteable address|all relevant MX records point to non-existent hosts)\s*$
            ^%(pid)s \w+ authenticator failed for (?:[^\[\( ]* )?(?:\(\S*\) )?\[<HOST>\](?::\d+)?(?: I=\[\S+\](:\d+)?)?: 535 Incorrect authentication data( \(set_id=.*\)|: \d+ Time\(s\))?\s*$
            ^%(pid)s %(host_info)srejected RCPT [^@]+@\S+: (?:relay not permitted|Sender verify failed|Unknown user|Unrouteable address)\s*$
            ^%(pid)s SMTP protocol synchronization error \([^)]*\): rejected (?:connection from|"\S+") %(host_info)s(?:next )?input=".*"\s*$
            ^%(pid)s SMTP call from (?:[^\[\( ]* )?%(host_info)sdropped: too many (?:nonmail commands|syntax or protocol errors) \(last (?:command )?was "[^"]*"\)\s*$
            ^%(pid)s SMTP protocol error in "[^"]+(?:"+[^"]*(?="))*?" %(host_info)sAUTH command used when not advertised\s*$
            ^%(pid)s no MAIL in SMTP connection from (?:[^\[\( ]* )?(?:\(\S*\) )?%(host_info)sD=\d\S*s(?: C=\S*)?\s*$
            ^%(pid)s (?:[\w\-]+ )?SMTP connection from (?:[^\[\( ]* )?(?:\(\S*\) )?%(host_info)sclosed by DROP in ACL\s*$
            <mdre-<mode>>

mdre-aggressive = ^%(pid)s no host name found for IP address <HOST>$
                  ^%(pid)s no IP address found for host \S+ \(during SMTP connection from \[<HOST>\]\)$

mdre-normal = 

# Parameter `mode` - `normal` or `aggressive`.
# Aggressive mode can be used to match flood and ddos-similar log-entries like:
#   'no host found for IP', 'no IP found for host'.
# Note this is not an authentication failures, so it may produce lots of false 
# positives on misconfigured MTAs.
# Ex.:
#   filter = exim[mode=aggressive]
mode = normal

ignoreregex = 

# DEV Notes:
# The %(host_info) defination contains a <HOST> match
#
# SMTP protocol synchronization error \([^)]*\)  <- This needs to be non-greedy
# to void capture beyond ")" to avoid a DoS Injection vulnerabilty as input= is
# user injectable data.
#
# Author: Cyril Jaquier
#         Daniel Black (rewrote with strong regexs)
#         Martin O'Neal (added additional regexs to detect authentication failures, protocol errors, and drops) 


```
What's wrong?

**Comment by ysha0** [comment]
How to correct filter for catch line beginning:

> 2023-04-01 07:34:55 SMTP call from [164.52.0.102] dropped: too many syntax or protocol errors (last command was
> 2023-04-01 07:35:18 SMTP call from [164.52.0.102] dropped: too many unrecognized commands (last was
>

**Reply by sebres** [reply]
You don't need to correct filter - it is enough to add additional failregex directly to the jail (in jail.local where you enable it and set other params).
```ini
[exim]
failregex = %(known/failregex)s
            ^\s*SMTP call from \[<ADDR>\] dropped: too many\b
enabled = true
```

**Comment by ysha0** [comment]
Thanks, how can I test the exim filter with added failregex?

As usual `fail2ban-regex /var/log/exim/mainlog /usr/local/etc/fail2ban/filter.d/exim.conf` or should I use another way?

**Reply by sebres** [reply]
At the moment you can't test jail with fail2ban-regex.
So you have to add this RE to the filter (e. g. like this to `filter.d/exim.local`, note **.local** not **.conf**):
```ini
[Definition]
failregex = %(known/failregex)s
            ^\s*SMTP call from \[<ADDR>\] dropped: too many\b
```
then you'd be able to test it using `fail2ban-regex /var/log/exim/mainlog exim`

But you could test the RE directly:
```bash
# single RE:
fail2ban-regex /var/log/exim/mainlog '^\s*SMTP call from \[<ADDR>\] dropped: too many\b'
# or combined with the filter's failregex:
fail2ban-regex /var/log/exim/mainlog "`printf 'exim[failregex="<known/failregex>\n%s"]' '^\s*SMTP call from \[<ADDR>\] dropped: too many\b'`"
```

**Comment by ysha0** [comment]
Hi, thanks

First method working Ok with cmd:
fail2ban-regex /var/log/exim/mainlog '^\s*SMTP call from \[<ADDR>\] dropped: too many\b'

But second method with combinations filters (known+new) returns an error:
```
fail2ban-regex /var/log/exim/mainlog "`printf 'exim[failregex="<known/failregex>\n%s"]' '^\s*SMTP call from \[<ADDR>\] dropped: too many\b'`"
known/failregex: No such file or directory.
```
Somesing wrong, looks like syntax error. How to correct second method for check full filter?

**Reply by sebres** [reply]
Dunno. It is depending on shell (I wrote it for bash)... Guess one'd escape \<\> chars somehow to avoid redirect in your preferred shell.
If you're not familiar with that, just create local filter as described above.

**Comment by ysha0** [comment]
Yes, in bash second method also work. Initially I ran both methods in sh. Thanks for clarifying about "bash".

**Comment by ysha0** [comment]
Hi, how may I add once more line in parameter failregex?

so it was:
```
[Definition]
failregex = %(known/failregex)s
            ^\s*SMTP call from \[<ADDR>\] dropped: too many\b

```
Line from log:
`2023-04-20 13:22:57 unexpected disconnection while reading SMTP command from (2284cd7572bb919e2d87ce6580644e47be43e2) [222.160.142.218] D=5s`

after add another failregex condition:
```
[Definition]
failregex = %(known/failregex)s
            ^\s*SMTP call from \[<ADDR>\] dropped: too many\b
            ^\s*unexpected disconnection while reading SMTP command from \[<ADDR>\]\b

```

Is this right?

**Reply by sebres** [reply]
The way how you add additional RE is correct (after new-line and at least one space)...

But your RE is not correct:
- there is still some session ID enclosed in round parenthesis before the IP, either use `\(\S+\)\s` or better something like `(?:[^\[]+ )?` (optional, ignores anything before open square bracket)
- `\b` after `\]` would not work (it is not a word to use word-boundary), either use space `\s` or simply remove it (unneeded because closed bracket is unique, `\b` normally used to signal end of word, e. g. `many\b` to match `many` but not `manyame`, `manyone` etc).

Something like that would be correct and match your log line:
```diff
- ^\s*unexpected disconnection while reading SMTP command from \[<ADDR>\]\b
+ ^\s*unexpected disconnection while reading SMTP command from (?:[^\[]+ )?\[<ADDR>\]
```

**Comment by ysha0** [comment]
Hi,

Line from log:
`2023-04-24 15:00:43 SMTP call from impressive.monitoring.internet-measurement.com [87.236.176.70] dropped: too many syntax or protocol errors (last command was "\300\024?\234?\235?/?5\300\022?",  C=EHLO)`

Why not worked filter.d/exim.conf ?
`^%(pid)s SMTP call from (?:[^\[\( ]* )?%(host_info)sdropped: too many (?:nonmail commands|syntax or protocol errors) \(last (?:command )?was "[^"]*"\)\s*$`

What RE using for this case?

**Reply by sebres** [reply]
There is a `,  C=EHLO)` after closed quote, so `"\)` would obviously not match.
One doesn't really need the whole match expression and right anchor, so something like this would be enough:
```diff
- ^%(pid)s SMTP call from (?:[^\[\( ]* )?%(host_info)sdropped: too many (?:nonmail commands|syntax or protocol errors) \(last (?:command )?was "[^"]*"\)\s*$
+ ^%(pid)s SMTP call from (?:[^\[\( ]* )?%(host_info)sdropped: too many (?:nonmail commands|syntax or protocol errors)
```

**Comment by ysha0** [comment]
working config jail.local
```
before = paths-freebsd.conf

[exim]
failregex = %(known/failregex)s
            ^[^\[]*\[<ADDR>\] F=(?:<[^>]*>|[^@]+@\S+) rejected RCPT [^@]+@\S+: (?:Your IP in HELO|Invalid sender address)\b
            ^\s*SMTP call from \[<ADDR>\] dropped: too many\b
```

after adding new regex line:
```
[exim]
failregex = %(known/failregex)s
            ^[^\[]*\[<ADDR>\] F=(?:<[^>]*>|[^@]+@\S+) rejected RCPT [^@]+@\S+: (?:Your IP in HELO|Invalid sender address)\b
            ^%(pid)s SMTP call from (?:[^\[\( ]* )?%(host_info)sdropped: too many (?:nonmail commands|syntax or protocol errors)
            
```
Error occurs when start fail2ban:

2023-04-24 20:03:41,866 fail2ban                [88729]: ERROR   Failed during configuration: Bad value substitution: option 'failregex' in section 'exim' contains an interpolation key 'pid' which is not a valid option name. Raw value: '%(known/failregex)s\n^[^\\[]*\\[<ADDR>\\] F=(?:<[^>]*>|[^@]+@\\S+) rejected RCPT [^@]+@\\S+: (?:Your IP in HELO|Invalid sender address)\\b\n^%(pid)s SMTP call from (?:[^\\[\\( ]* )?%(host_info)sdropped: too many (?:nonmail commands|syntax or protocol errors)\n^\\s*SMTP call from \\[<ADDR>\\] dropped: too many\\b'

**Reply by sebres** [reply]
Yes, `%(pid)s` as well as `%(host_info)s` are substituting variables in jail (that are not available there).
Either remove `%(pid)s` (you seem not to have it in log) and replace `%(host_info)s` with related RE from exim-common. |
Or better add REs to filter (`filter.d/exim.local`) in similar way (but to the definition section), however I'd change it directly in `filter.d/exim.conf` (I tend to rewrite this RE for next version).

**Comment by ysha0** [comment]
> Or better add REs to filter (filter.d/exim.local) in similar way (but to the definition section), however I'd change it directly in filter.d/exim.conf (I tend to rewrite this RE for next version).

Could you write how to change the config more correctly, please, I'm not a Fail2ban expert yet :(
```
[exim]
failregex = %(known/failregex)s
            ^[^\[]*\[<ADDR>\] F=(?:<[^>]*>|[^@]+@\S+) rejected RCPT [^@]+@\S+: (?:Your IP in HELO|Invalid sender address)\b
            ^%(pid)s SMTP call from (?:[^\[\( ]* )?%(host_info)sdropped: too many (?:nonmail commands|syntax or protocol errors)
            ^\s*SMTP call from \[<ADDR>\] dropped: too many\b

```

**Reply by sebres** [reply]
/etc/fail2ban/filter.d/exim.local
```ini
[Definition]
failregex = %(known/failregex)s
            ^%(pid)s SMTP call from (?:[^\[\( ]* )?%(host_info)sdropped: too many (?:nonmail commands|syntax or protocol errors)
```
if you'd add all your custom REs here, you can remove it from jail.local

**Comment by ysha0** [comment]
Thanks, It's working

**Comment by ysha0** [comment]
Hi, which failregex use for:

```
2023-06-02 13:09:04 H=([117.32.132.170]) [117.32.132.170] F=<6zie7zjt9sm7v@opticool.cz> rejected RCPT <faturamento@yyy.zz>: HELO blacklisted
2023-06-02 18:26:00 H=([95.140.209.25]) [95.140.209.25] F=<tara@yyy.zz> rejected RCPT <tara@yyy.zz>: HELO blacklisted
```
This not worked:

`^[^\[]*\[<ADDR>\] F=(?:<[^>]*>|[^@]+@\S+) rejected RCPT [^@]+@\S+: (?:Your IP in HELO|Invalid sender address)\b
`

**Reply by sebres** [reply]
It doesn't work for two reasons:
- `[...]` in `H=([...])` gets mistakenly matched by `\[<ADDR>\]`, however it must match second `[...]`
- obviously because the reject reason `HELO blacklisted` isn't in the RE

Something like that would match it:
```diff
- ^[^\[]*\[<ADDR>\] F=(?:<[^>]*>|[^@]+@\S+) rejected RCPT [^@]+@\S+: (?:Your IP in HELO|Invalid sender address)\b
+ ^[^\[]* (?:H=([\w.-]+ )?(?:\(\S+\) )?)?\[<ADDR>\] F=(?:<[^>]*>|[^@]+@\S+) rejected RCPT [^@]+@\S+: (?:Your IP in HELO|Invalid sender address|HELO blacklisted)\b
```
for current filter this line need to be extended (with `HELO blacklisted`):
https://github.com/fail2ban/fail2ban/blob/48c91dfb6bd8478968c226d0c6ca965bdce2768d/config/filter.d/exim.conf#L21
or even for every rejected RCPT:
```diff
-             ^%(pid)s %(host_info)srejected RCPT [^@]+@\S+: (?:relay not permitted|Sender verify failed|Unknown user|Unrouteable address)\s*$
+             ^%(pid)s %(host_info)srejected RCPT\b

```

**Comment by ysha0** [comment]
Chose the simplest variant for "every rejected RCP" and it works, thanks:

`^%(pid)s %(host_info)srejected RCPT\b`
